vec.ecl

```diff
@@ -0,0 +1,137 @@
+/*##############################################################################
+## HPCC SYSTEMS software Copyright (C) 2020 HPCC Systems.  All rights reserved.
+############################################################################## */
+IMPORT $ as PBBlas;
+/**
+  *  This module handles different vector (column matrix) operation
+  */
+EXPORT vec := MODULE
+ 
+ 
+ /**
+  *  Creation of a vector with given length and constant value.
+  *
+  *  @param N      Length of the vector in UNSIGNED.
+  *  @param def    Constant value of the vector in REAL.
+  *  @return       Matrix (vector) in DATASET(Layout_Cell) form.
+  *  @see          PBblas/Types.Layout_Cell
+  */
+ EXPORT From(UNSIGNED N,REAL8 def = 1.0) := FUNCTION
+  seed := DATASET([{1,0,0,0}], PBblas.Types.Layout_cell);
+  PerCluster := ROUNDUP(N/CLUSTERSIZE);
+  PBblas.Types.Layout_cell addNodeNum(PBblas.Types.Layout_cell L, UNSIGNED4 c) := TRANSFORM
+   SELF.wi_id := L.wi_id;
+   SELF.x := (c-1)*PerCluster;
+   SELF.y := 1;
+   SELF.v := def;
+ END;
+ // Create eventual vector across all nodes (in case it is huge)
+ one_per_node := DISTRIBUTE(NORMALIZE(seed, CLUSTERSIZE, addNodeNum(LEFT, COUNTER)), x DIV PerCluster);
+ PBblas.Types.Layout_cell fillRow(PBblas.Types.Layout_cell L, UNSIGNED4 c) := TRANSFORM
+  SELF.wi_id := L.wi_id;
+  SELF.x := L.x + c;
+  SELF.y := 1;
+  SELF.v := def;
+ END;
+ // Now generate on each node; filter 'clips' of the possible extra '1' generated on some nodes
+ m := NORMALIZE(one_per_node, PerCluster, fillRow(LEFT,COUNTER))(x <= N);
+ RETURN m;
+ END;
+ 
+ 
+  /**
+  *  Creation of a diagonal matrix from a given vector.
+  *
+  *  @param v      Given vector in DATASET(Layout_cell).
+  *  @param N      Dimension of the diagonal matrix in UNSIGNED 
+  *  @return       Matrix in DATASET(Layout_Cell) form.
+  *  @see          PBblas/Types.Layout_Cell
+  */
+  
+  
+ EXPORT ToDiag(DATASET(PBblas.Types.Layout_cell) v, UNSIGNED N=1) := 
+  PROJECT(v(x>=N), TRANSFORM(PBblas.Types.Layout_Cell, 
```

Comment:
Is this code correct?  I was expecting x <= N and I'm not sure what the -(N-1) term did either.  Please can you explain.

Reply:
@ghalliday Wow.  This was puzzling.  It passed the tests and did the right thing, but the documentation on the meaning of the  parameter N was wrong.  He wrote this to be the complement to FromDiag, and the N really relates to K on the FromDiag.  The tests run with N = 1 (default), and it actually creates a diag matrix, but the N parameter is useless (and not used), as far as I can tell.  The size of the matrix is implicit from the size of the vector.  If N was set to 2, it would create a vecLen - 1 square matrix, ignoring the first component.  I'm removing the parameter N and simplifying.

Reply:
@ghalliday I'm also removing parameter K from FromDiag().  It is confusing, untested, and unused.